File: 面试高频题/0015 三数之和.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if len(nums) < 3: return []
        res = []
        nums.sort()
        for i in range(0, len(nums)-2):
            if nums[i] > 0: break
            if i>0 and nums[i]==nums[i-1]:
                continue
            j = i + 1
            k = len(nums) - 1
            while j < k:
                sum_abc = nums[i] + nums[j] + nums[k]
                if sum_abc < 0:
                    j += 1
                    while j<k and nums[j]==nums[j-1]:
                        j += 1
                elif sum_abc > 0:
                    k -= 1
                    while j<k and nums[k]==nums[k+1]:
                        k -= 1
                else:
                    res.append([nums[i], nums[j], nums[k]])
                    j += 1
                    k -= 1
                    while j<k and nums[j]==nums[j-1]:
                        j += 1
                    while j<k and nums[k]==nums[k+1]:
                        k -= 1
        return res

File: 面试高频题/0015 三数之和.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        if (n < 3)
            return {};
        vector<vector<int>> res;
        int L, R;
        int sum;
        for (int i=0; i<n-2; ++i) {
            if (nums[i] > 0)
                break;
            if (i>0 && nums[i]==nums[i-1])
                continue;
            L = i+1;
            R = n-1;
            while (L<R) {
                sum = nums[i] + nums[L] + nums[R];
                if (sum<0) {
                    ++L;
                    while (L<R && nums[L]==nums[L-1])
                        ++L;
                } else if (sum>0) {
                    --R;
                    while (L<R && nums[R]==nums[R+1])
                        --R;
                } else {
                    res.push_back({nums[i], nums[L], nums[R]});
                    ++L;
                    --R;
                    while (L<R && nums[L]==nums[L-1])
                        ++L;
                    while (L<R && nums[R]==nums[R+1])
                        --R;
                }
            }
        }
        return res;
    }
};